In [1]:
# imports
import pandas as pd
import sys

In [2]:
# import custom libraries and scripts
# sys.path.append(os.path.abspath(os.path.join("../..")))
sys.path.append(".")
sys.path.append("..")

from defaults import *
from extract_dataframe import read_json
from extract_dataframe import TweetDfExtractor
from clean_tweets_dataframe import Clean_Tweets

In [3]:
# read data
global_data = pd.read_json('../data/processed_global_tweet_data.json')
global_data.head(10)

created_at  \
0 2022-08-07 22:31:20+00:00   
1 2022-08-07 22:31:16+00:00   
2 2022-08-07 22:31:07+00:00   
3 2022-08-07 22:31:06+00:00   
4 2022-08-07 22:31:04+00:00   
5 2022-08-07 22:31:02+00:00   
6 2022-08-07 22:30:59+00:00   
7 2022-08-07 22:30:59+00:00   
8 2022-08-07 22:30:50+00:00   
9 2022-08-07 22:30:45+00:00   

                                              source  \
0  <a href="http://twitter.com/download/android" ...   
1  <a href="http://twitter.com/download/android" ...   
2  <a href="http://twitter.com/download/android" ...   
3  <a href="http://twitter.com/download/android" ...   
4  <a href="http://twitter.com/download/iphone" r...   
5  <a href="http://twitter.com/download/android" ...   
6  <a href="http://twitter.com/download/android" ...   
7  <a href="http://twitter.com/download/android" ...   
8  <a href="http://twitter.com/download/android" ...   
9  <a href="https://mobile.twitter.com" rel="nofo...   

                                       original_text      polarity  \
0  RT @i_ameztoy: Extra random image (I):\n\nLets... -1.250000e-01   
1  RT @IndoPac_Info: #China's media explains the ... -1.000000e-01   
2  China even cut off communication, they don't a...  0.000000e+00   
3  Putin to #XiJinping : I told you my friend, Ta...  1.000000e-01   
4  RT @ChinaUncensored: I’m sorry, I thought Taiw... -6.938894e-18   
5  RT @benedictrogers: We must not let this happe...  2.000000e-01   
6  RT @TGTM_Official: What kind of country can co...  1.583333e-01   
7  RT @ChinaInfo777: #PinkFloyd singer Roger Wate...  0.000000e+00   
8  RT @AmbQinGang: China's SC&amp;FM Wang Yi elab...  0.000000e+00   
9  RT @CGMeifangZhang: Chinese ambassador to the ...  2.000000e-01   

   subjectivity lang  favorite_count  status_count  retweet_count  \
0      0.190625   en               4          8097              2   
1      0.100000   en             691          5831            201   
2      0.000000   en               0          1627              0   
3      0.350000   en               0          1627              0   
4      0.556250   en            1521         18958            381   
5      0.500000   en             116         48483             36   
6      0.800000   en            1106          4173            411   
7      0.000000   en              10         24102              5   
8      0.000000   en            1221           630            239   
9      0.375000   en              49        107188             25   

       screen_name  original_author  followers_count  friends_count  \
0        i_ameztoy        i_ameztoy            20497           2621   
1            ZIisq            ZIisq               65            272   
2        Fin21Free        Fin21Free               85            392   
3        Fin21Free        Fin21Free               85            392   
4   VizziniDolores   VizziniDolores              910           2608   
5  GraceCh15554845  GraceCh15554845              207             54   
6      Philipkuma1      Philipkuma1               12            264   
7        nhohn2011        nhohn2011              870            508   
8  ClaudioColomaRI  ClaudioColomaRI              127            263   
9        jmarzola1        jmarzola1              213            877   

   possibly_sensitive                                           hashtags  \
0                 NaN          [{'text': 'City', 'indices': [132, 137]}]   
1                 NaN  [{'text': 'China', 'indices': [18, 24]}, {'tex...   
2                 NaN     [{'text': 'XiJinping', 'indices': [127, 137]}]   
3                 NaN        [{'text': 'XiJinping', 'indices': [9, 19]}]   
4                 NaN                                                 []   
5                 0.0          [{'text': 'Taiwan', 'indices': [84, 91]}]   
6                 NaN  [{'text': 'Taiwan', 'indices': [101, 108]}, {'...   
7                 NaN  [{'text': 'PinkFloyd', 'indices': [18, 28]}, {...   
8                 NaN          [{'text': 'Taiwan', 'indices':

In [4]:
global_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22000 entries, 0 to 21999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   created_at          22000 non-null  datetime64[ns, UTC]
 1   source              22000 non-null  object             
 2   original_text       22000 non-null  object             
 3   polarity            22000 non-null  float64            
 4   subjectivity        22000 non-null  float64            
 5   lang                22000 non-null  object             
 6   favorite_count      22000 non-null  int64              
 7   status_count        22000 non-null  int64              
 8   retweet_count       22000 non-null  int64              
 9   screen_name         22000 non-null  object             
 10  original_author     22000 non-null  object             
 11  followers_count     22000 non-null  int64              
 12  friends_count       22000 non-nu

In [5]:
global_data.describe()

polarity  subjectivity  favorite_count  status_count  \
count  22000.000000  22000.000000    22000.000000  2.200000e+04   
mean       0.061325      0.283839      852.137318  5.446036e+04   
std        0.223701      0.290963     3106.077645  1.454120e+05   
min       -1.000000      0.000000        0.000000  1.000000e+00   
25%        0.000000      0.000000        2.000000  2.105750e+03   
50%        0.000000      0.200000      115.000000  1.038750e+04   
75%        0.133333      0.468824      655.000000  4.526150e+04   
max        1.000000      1.000000    65170.000000  4.108317e+06   

       retweet_count  followers_count  friends_count  possibly_sensitive  
count   22000.000000     2.200000e+04   22000.000000         6191.000000  
mean      176.750182     1.796764e+04    1563.114455            0.037151  
std       498.435765     3.030478e+05    4358.651264            0.189146  
min         0.000000     0.000000e+00       0.000000            0.000000  
25%         2.000000     5.700000e+01     137.000000            0.000000  
50%        38.000000     2.840000e+02     487.000000            0.000000  
75%       187.000000     1.324500e+03    1599.000000            0.000000  
max     17409.000000     1.449852e+07  208360.000000            1.000000

## EDA

### Remove duplicated rows

In [6]:
clean_tweets = Clean_Tweets(global_data)
clean_tweets.df.shape

Automation in Action...!!!


(22000, 17)

In [7]:
clean_tweets.df.shape

(22000, 17)

In [8]:
clean_tweets = clean_tweets.drop_duplicate(global_data)
clean_tweets.shape

(7440, 17)

As we can see there were too many duplicates

### Remove tweets that are not english

In [9]:
clean_tweets.lang.value_counts()

en    7440
Name: lang, dtype: int64

All tweets are in English

### Missing values

In [10]:
clean_tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7440 entries, 0 to 21997
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   created_at          7440 non-null   datetime64[ns, UTC]
 1   source              7440 non-null   object             
 2   original_text       7440 non-null   object             
 3   polarity            7440 non-null   float64            
 4   subjectivity        7440 non-null   float64            
 5   lang                7440 non-null   object             
 6   favorite_count      7440 non-null   int64              
 7   status_count        7440 non-null   int64              
 8   retweet_count       7440 non-null   int64              
 9   screen_name         7440 non-null   object             
 10  original_author     7440 non-null   object             
 11  followers_count     7440 non-null   int64              
 12  friends_count       7440 non-null

Only the feature possibility sensitive have a missing value

In [11]:
clean_tweets.possibly_sensitive.value_counts()

0.0    3866
1.0     111
Name: possibly_sensitive, dtype: int64

As we can see there are 3,866 not sensitive and 111 sensitive tweets
And only 3,988 out of 77,000 tweets are recorded for sensitivity

### Handling missing values

In [12]:
clean_tweets = Clean_Tweets.fill_missing(clean_tweets, df = clean_tweets, column="possibly_sensitive", value = "unknown")
clean_tweets.isna().sum()

created_at            0
source                0
original_text         0
polarity              0
subjectivity          0
lang                  0
favorite_count        0
status_count          0
retweet_count         0
screen_name           0
original_author       0
followers_count       0
friends_count         0
possibly_sensitive    0
hashtags              0
user_mentions         0
place                 0
dtype: int64

In [13]:
clean_tweets.columns

Index(['created_at', 'source', 'original_text', 'polarity', 'subjectivity',
       'lang', 'favorite_count', 'status_count', 'retweet_count',
       'screen_name', 'original_author', 'followers_count', 'friends_count',
       'possibly_sensitive', 'hashtags', 'user_mentions', 'place'],
      dtype='object')

In [14]:
clean_tweets[['retweet_count', 'source', 'original_text', 'hashtags', 'place']]

retweet_count                                             source  \
0                  2  <a href="http://twitter.com/download/android" ...   
1                201  <a href="http://twitter.com/download/android" ...   
2                  0  <a href="http://twitter.com/download/android" ...   
3                  0  <a href="http://twitter.com/download/android" ...   
4                381  <a href="http://twitter.com/download/iphone" r...   
...              ...                                                ...   
21974              3  <a href="https://mobile.twitter.com" rel="nofo...   
21987              1  <a href="http://twitter.com/download/iphone" r...   
21989              0  <a href="http://twitter.com/download/iphone" r...   
21991              0  <a href="http://twitter.com/download/android" ...   
21997              0  <a href="http://twitter.com/download/android" ...   

                                           original_text  \
0      RT @i_ameztoy: Extra random image (I):\n\nLets...   
1      RT @IndoPac_Info: #China's media explains the ...   
2      China even cut off communication, they don't a...   
3      Putin to #XiJinping : I told you my friend, Ta...   
4      RT @ChinaUncensored: I’m sorry, I thought Taiw...   
...                                                  ...   
21974  RT @metesohtaoglu: 📌📸 Map of #China's possible...   
21987  RT @NEVERBOW: China is doing #exactly what #Ru...   
21989  Minister Wu is crystal clear in his @BBCNews i...   
21991  Reports say that #China is planning to seize #...   
21997                @Reuters Thanks #Pelosi smart move.   

                                                hashtags             place  
0              [{'text': 'City', 'indices': [132, 137]}]                    
1      [{'text': 'China', 'indices': [18, 24]}, {'tex...                    
2         [{'text': 'XiJinping', 'indices': [127, 137]}]       Netherlands  
3            [{'text': 'XiJinping', 'indices': [9, 19]}]       Netherlands  
4                                                     []    Ayent, Schweiz  
...                                                  ...               ...  
21974  [{'text': 'China', 'indices': [29, 35]}, {'tex...       Seattle, WA  
21987  [{'text': 'exactly', 'indices': [29, 37]}, {'t...                    
21989  [{'text': 'Taiwan', 'indices': [168, 175]}, {'...   Toronto, Canada  
21991  [{'text': 'China', 'indices': [17, 23]}, {'tex...                    
21997          [{'text': 'Pelosi', 'indices': [16, 23]}]  🇺🇲🇷🇺🇺🇦🇫🇷🇦🇪🇮🇱🏳️‍🌈  

[7440 rows x 5 columns]

There are empty place values

### Replace empty strings

In [15]:
clean_tweets = Clean_Tweets.replace_empty_string(clean_tweets, df = clean_tweets, column="place", value = "unknown")
clean_tweets['place'].value_counts()

unknown                           2805
Việt Nam                           116
India                              107
United States                       72
Turn on 🔔                           57
                                  ... 
New York, New York                   1
Fontaines-Saint-Martin, France       1
🇺🇲🇺🇲🇺🇲                               1
Lisbon                               1
🇺🇲🇷🇺🇺🇦🇫🇷🇦🇪🇮🇱🏳️‍🌈                     1
Name: place, Length: 1809, dtype: int64

### Convert to datetime

In [16]:
clean_tweets['created_at'] = pd.to_datetime(clean_tweets['created_at'])
clean_tweets.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7440 entries, 0 to 21997
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   created_at          7440 non-null   datetime64[ns, UTC]
 1   source              7440 non-null   object             
 2   original_text       7440 non-null   object             
 3   polarity            7440 non-null   float64            
 4   subjectivity        7440 non-null   float64            
 5   lang                7440 non-null   object             
 6   favorite_count      7440 non-null   int64              
 7   status_count        7440 non-null   int64              
 8   retweet_count       7440 non-null   int64              
 9   screen_name         7440 non-null   object             
 10  original_author     7440 non-null   object             
 11  followers_count     7440 non-null   int64              
 12  friends_count       7440 non-null

### extract source of tweets

In [17]:
#clean_tweets["source"] = clean_tweets["source"].apply(Clean_Tweets.extract_device_name(self = clean_tweets, source='source'))

### save current dataset

In [18]:
clean_tweets.to_csv('../data/clean_data.csv', index = False)
print('clean data saved successfully')

clean data saved successfully
